interop-tests/src/test/java/example/myapp/helloworld/grpc/RichErrorModelNativeTest.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2018-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package example.myapp.helloworld.grpc;
+
+import akka.actor.ActorSystem;
+import akka.grpc.AkkaGrpcException;
+import akka.grpc.GrpcClientSettings;
+import akka.http.javadsl.Http;
+import akka.http.javadsl.ServerBinding;
+import akka.http.javadsl.model.HttpRequest;
+import akka.http.javadsl.model.HttpResponse;
+import com.google.rpc.error_details.LocalizedMessage;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+import io.grpc.Status;
+import org.junit.Assert;
+import org.junit.Test;
+import org.scalatestplus.junit.JUnitSuite;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import static org.junit.Assert.assertEquals;
+
+
+public class RichErrorModelNativeTest extends JUnitSuite {
+
+    private ServerBinding run(ActorSystem sys) throws Exception {
+
+        GreeterService impl = new RichErrorNativeImpl();
+
+        akka.japi.function.Function<HttpRequest, CompletionStage<HttpResponse>> service = GreeterServiceHandlerFactory.create(impl, sys);
+        CompletionStage<ServerBinding> bound = Http
+                .get(sys)
+                .newServerAt("127.0.0.1", 8091)
+                .bind(service);
+
+        bound.thenAccept(binding -> {
+            System.out.println("gRPC server bound to: " + binding.localAddress());
+        });
+        return bound.toCompletableFuture().get();
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testNativeApi() throws Exception {
+        Config conf = ConfigFactory.load();
+        ActorSystem sys = ActorSystem.create("HelloWorld", conf);
+        run(sys);
+
+        GrpcClientSettings settings = GrpcClientSettings.connectToServiceAt("127.0.0.1", 8091, sys).withTls(false);
+
+        GreeterServiceClient client = null;
+        try {
+            client = GreeterServiceClient.create(settings, sys);
+
+            // #client_request
+            HelloRequest request = HelloRequest.newBuilder().setName("Alice").build();
+            CompletionStage<HelloReply> response = client.sayHello(request);
+            AkkaGrpcException statusEx = response.toCompletableFuture().handle((res, ex) -> {
+                return (AkkaGrpcException) ex;
+            }).get();
+
+            assertEquals("type.googleapis.com/google.rpc.LocalizedMessage", statusEx.status().getDetails(0).getTypeUrl());
+
+            assertEquals(Status.INVALID_ARGUMENT.getCode().value(), statusEx.code());
+            assertEquals("What is wrong?", statusEx.message());
+
+            LocalizedMessage details = statusEx.getDetails(0, com.google.rpc.error_details.LocalizedMessage.messageCompanion());
```

Comment:
I'm not too sold on this API: it's still pretty verbose, and you still need to first check the `typeUrl` to see the type before you can call `getDetail` with the correct `messageCompanion` to parse it.

I wonder if we should generalize the `AnySupport` helper class we use in the akkaserverless-java-sdk (https://github.com/lightbend/akkaserverless-java-sdk/blob/main/sdk/java-sdk/src/main/scala/com/akkaserverless/javasdk/impl/AnySupport.scala), though turning that into a public API would probably require quite some API design effort. I wonder if there's any conventions around this on the grpc-java side yet?

That way, you could just use the regular `status.getDetailsList.asScala`, and turn them into domain classes with that more general-purpose utility.

As a less ambitious first step, we could provide a `GrpcServiceExtension.getDetails(status)` that just supports the types from https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto . A question is what we should do with protobuf objects we don't recognize - I guess we could just return them in their `Any` form.

Reply:
I am still looking into this.